Observer/RefundOrderInventoryObserver.php

```diff
@@ -0,0 +1,200 @@
+        \Magento\Catalog\Model\Indexer\Product\Price\Processor $priceIndexer,
+        \Magento\SalesInventory\Model\Order\ReturnProcessor $returnProcessor,
+        \Magento\Sales\Api\OrderRepositoryInterface $orderRepository,
+        WebsiteRepositoryInterface $websiteRepository,
+        SalesEventInterfaceFactory $salesEventFactory,
+        SalesChannelInterfaceFactory $salesChannelFactory,
+        GetSourceSelectionResultFromOrder $getSourceSelectionResultFromOrder,
+        ItemToDeductFactory $itemToDeductFactory,
+        SourceDeductionRequestFactory $sourceDeductionRequestFactory,
+        SourceDeductionService $sourceDeductionService
+    ) {
+        $this->stockConfiguration = $stockConfiguration;
+        $this->stockManagement = $stockManagement;
+        $this->stockIndexerProcessor = $stockIndexerProcessor;
+        $this->priceIndexer = $priceIndexer;
+        $this->returnProcessor = $returnProcessor;
+        $this->orderRepository = $orderRepository;
+        $this->websiteRepository = $websiteRepository;
+        $this->salesEventFactory = $salesEventFactory;
+        $this->salesChannelFactory = $salesChannelFactory;
+        $this->getSourceSelectionResultFromOrder = $getSourceSelectionResultFromOrder;
+        $this->itemToDeductFactory = $itemToDeductFactory;
+        $this->sourceDeductionRequestFactory = $sourceDeductionRequestFactory;
+        $this->sourceDeductionService = $sourceDeductionService;
+    }
+
+    /**
+     * Return creditmemo items qty to stock
+     *
+     * @param EventObserver $observer
+     * @return void
+     */
+    public function execute(EventObserver $observer)
+    {
+        /* @var $creditmemo \Magento\Sales\Model\Order\Creditmemo */
+        $creditmemo = $observer->getEvent()->getCreditmemo();
+        $order = $this->orderRepository->get($creditmemo->getOrderId());
+        $returnToStockItems = [];
+
+        $websiteId = $order->getStore()->getWebsiteId();
+        $websiteCode = $this->websiteRepository->getById($websiteId)->getCode();
+        $salesChannel = $this->salesChannelFactory->create([
+            'data' => [
+                'type' => SalesChannelInterface::TYPE_WEBSITE,
+                'code' => $websiteCode
+            ]
+        ]);
+
+        $salesEvent = $this->salesEventFactory->create([
+            'type' => SalesEventInterface::EVENT_CREDITMEMO_CREATED,
+            'objectType' => SalesEventInterface::OBJECT_TYPE_ORDER,
+            'objectId' => (string)$creditmemo->getOrderId()
+        ]);
+
+        $sourceSelectionItems = $this->getSourceSelectionResultFromOrder->execute($order)->getSourceSelectionItems();
+
+        foreach ($creditmemo->getItems() as $item) {
+            if ($item->getBackToStock()) {
+                $returnToStockItems[] = $item->getOrderItemId();
+
+                foreach ($sourceSelectionItems as $sourceSelectionItem) {
+                    if ($sourceSelectionItem->getSku() === $item->getSku()) {
```

Comment:
The nesting of this functionality is too deep really, you'd see issues thrown about the cyclomatic complexity of this function if you passed it through a linter